testing-daml-models/testing-with-scenarios/5_scenario_tracing.md

```diff
@@ -0,0 +1,40 @@
+In this final step, you will learn how to add assertions and good old `printf` debugging to DAML
+scenarios.
+
+First, we'll have `Alice` check that she now really follows `Bob`:
+
+<pre class="file" data-filename="daml/User.daml" data-target="append">
+    submit alice do 
+      aliceUser <- fetch newAliceCid
+      assertMsg "Whoops!" (aliceUser.following == [bob])
+</pre>
+
+The `assertMsg` takes a text that is printed in case the assertion doesn't hold and a boolean
+expression. If you replace the last line with an assertion that doesn't hold, you'll see the
+error message in the scenario results.
+
+<pre class="file" data-filename="daml/User.daml" data-target="append">
+      assertMsg "Whoops!" $ following == [alice]
+</pre>
+
+Let's remove the failing line again.
+
+Once your scenarios grow bigger, it's helpful to trace the execution flow of your scenario. With the
+function `trace : Text -> a -> a` you can add tracing information to any expression, not just
+scenario expressions. For example, you can add a trace just before a new `User` contract is created
+after the `Follow` choice by replacing the line 
+
+```
+create this with following = userToFollow :: following
+```
+
+with
+
+<pre class="file" data-filename="daml/User.daml" data-target="clipboard">
+      trace "adding new follower" (create this with following = userToFollow :: following)
+</pre>
```

Comment:
Tripple backticks and the <pre> tag don't really play well together. I get two buttons for copying the trace code and when I click on the button it copies the whole text

![image](https://user-images.githubusercontent.com/56260209/80570219-b7668500-89fa-11ea-95d8-c1b7e835133d.png)